### PR TITLE
Add Wav2Vec2 & Hubert ForSequenceClassification

### PR DESCRIPTION
**Pull request by anton-l**
# What does this PR do?

This adds a Hubert extension for sequence classification.
Ultimately this classification head should be compatible with s3prl `UtteranceLevel` [implementation](https://github.com/s3prl/s3prl/blob/master/s3prl/downstream/model.py#L35) to support classification tasks from SUPERB, such as [Keyword Spotting](https://huggingface.co/datasets/superb#ks) and transfer their pretrained models.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten @patil-suraj

### CODE REVIEW

File: src/transformers/models/hubert/configuration_hubert.py [hunk 1]
```diff
@@ -115,6 +115,11 @@ class HubertConfig(PretrainedConfig):
             Whether to zero infinite losses and the associated gradients of ``torch.nn.CTCLoss``. Infinite losses
             mainly occur when the inputs are too short to be aligned to the targets. Only relevant when training an
             instance of :class:`~transformers.HubertForCTC`.
+        use_weighted_layer_sum (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            Whether to use a weighted average of layer outputs with learned weights. Only relevant when training an
```
Comment (patrickvonplaten) [inline src/transformers/models/hubert/configuration_hubert.py:119]:
```suggestion
            Whether to use a weighted average of layer outputs with learned weights. Only relevant when using an
```
-> do the fine-tuned weights of `s3prl` have this set to ``True``?

Comment (anton-l) [reply src/transformers/models/hubert/configuration_hubert.py:119]: Yes, true for every task

File: src/transformers/models/hubert/modeling_hubert.py [hunk 2]
```diff
@@ -1070,3 +1071,123 @@ def forward(
...
+        if attention_mask is None:
+            pooled_output = hidden_states.mean(dim=1)
+        else:
+            output_lengths = self._get_feat_extract_output_lengths(attention_mask.sum(-1)).to(torch.long)
+            pooled_output = hidden_states.sum(dim=1) / output_lengths.view(-1, 1)
+
+        logits = self.classifier(pooled_output)
+
+        loss = None
+        if labels is not None:
```
Comment (patrickvonplaten) [inline src/transformers/models/hubert/modeling_hubert.py:1162]: Do we need all of them for now? I'm fine with it (that's also what we've done in `modeling_bert.py`, but if we don't need some losses for the superb benchmark, think it's better to not add them yet (always good to limit the functionality as much as possible in the beginning for better readability and less maintance)

File: tests/test_modeling_hubert.py [hunk 3]
```diff
@@ -579,3 +645,24 @@ def test_inference_ctc_batched(self):
             "sweat covered brion's body trickling into the tight loin cloth that was the only garment he wore",
         ]
         self.assertListEqual(predicted_trans, EXPECTED_TRANSCRIPTIONS)
+
+    def test_inference_classification_batched(self):
+        model = HubertForSequenceClassification.from_pretrained("anton-l/hubert-base-s3prl-superb-ks").to(torch_device)
```
Comment (patrickvonplaten) [inline tests/test_modeling_hubert.py:650]: Very cool!

File: tests/test_modeling_hubert.py [hunk 4]
```diff
@@ -579,3 +645,24 @@ def test_inference_ctc_batched(self):
...
+        processor = Wav2Vec2Processor.from_pretrained(
+            "anton-l/hubert-base-s3prl-superb-ks", do_lower_case=True, return_attention_mask=True
+        )
+        # TODO: use SUPERB Keyword Spotting instead
```
Comment (patrickvonplaten) [inline tests/test_modeling_hubert.py:654]: yes that would be nice! Feel free to create a dummy testing dataset for this one :-)

File: tests/test_modeling_hubert.py (hunk 3 above)
Comment (patrickvonplaten) [reply tests/test_modeling_hubert.py:650]: Let's try to add one test for each superb `ForClassification` problem

File: src/transformers/models/hubert/modeling_hubert.py [hunk 5]
```diff
@@ -20,12 +20,13 @@
...
 import torch.utils.checkpoint
 from torch import nn
+from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
 
 from transformers.deepspeed import is_deepspeed_zero3_enabled
 
 from ...activations import ACT2FN
 from ...file_utils import add_start_docstrings, add_start_docstrings_to_model_forward, replace_return_docstrings
-from ...modeling_outputs import BaseModelOutput, CausalLMOutput
+from ...modeling_outputs import BaseModelOutput, CausalLMOutput, SequenceClassifierOutput
```
Comment (patrickvonplaten) [inline src/transformers/models/hubert/modeling_hubert.py:29]: Great to reuse `SequenceClassifierOutput` here!

File: src/transformers/models/hubert/modeling_hubert.py (hunk 2 above)
Comment (anton-l) [reply src/transformers/models/hubert/modeling_hubert.py:1162]: Only single- and multilabel classification are relevant to SUPERB, I'll remove the other ones for now 👍

File: src/transformers/models/hubert/convert_hubert_original_s3prl_checkpoint_to_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    checkpoint = torch.load(checkpoint_path, map_location="cpu")
+    if checkpoint["Config"]["downstream_expert"]["modelrc"]["select"] not in SUPPORTED_MODELS:
+        raise NotImplementedError(f"The supported s3prl models are {SUPPORTED_MODELS}")
+
+    downstream_dict = checkpoint["Downstream"]
+
+    hf_congfig = HubertConfig.from_pretrained(config_path)
+    hf_model = HubertForSequenceClassification.from_pretrained(base_model_name, config=hf_congfig)
+    hf_feature_extractor = Wav2Vec2FeatureExtractor.from_pretrained(
+        base_model_name, return_attention_mask=True, do_normalize=False
```
Comment (patrickvonplaten) [inline src/transformers/models/hubert/convert_hubert_original_s3prl_checkpoint_to_pytorch.py:45]: You are sure none of the models are normalized? Are they usually based on this architecture: https://huggingface.co/facebook/hubert-base-ls960 ? E.g. `feat_extract_norm="group"` + `do_robust_layer_norm=False`? Like in: https://huggingface.co/facebook/hubert-base-ls960/blob/main/config.json#L43

File: src/transformers/models/hubert/modeling_hubert.py [hunk 7]
```diff
@@ -735,6 +736,18 @@ def _conv_out_length(input_length, kernel_size, stride):
 
         return input_lengths
 
+    def _get_feature_vector_attention_mask(self, feature_vector_length: int, attention_mask: torch.LongTensor):
```
Comment (patrickvonplaten) [inline src/transformers/models/hubert/modeling_hubert.py:739]: Awesome!

File: tests/test_modeling_hubert.py [hunk 8]
```diff
@@ -579,3 +656,102 @@ def test_inference_ctc_batched(self):
             "sweat covered brion's body trickling into the tight loin cloth that was the only garment he wore",
         ]
         self.assertListEqual(predicted_trans, EXPECTED_TRANSCRIPTIONS)
+
+    def test_inference_keyword_spotting(self):
```
Comment (patrickvonplaten) [inline tests/test_modeling_hubert.py:660]: Amazing tests!

File: tests/test_modeling_wav2vec2.py [hunk 9]
```diff
@@ -913,3 +927,25 @@ def test_loss_pretraining(self):
         expected_loss = 62.5170
 
         self.assertTrue(abs(outputs.loss.item() - expected_loss) < 1e-3)
+
+    def test_inference_emotion_recognition(self):
```
Comment (patrickvonplaten) [inline tests/test_modeling_wav2vec2.py:931]: maybe add other tasks here as well? To have 1-to-1 the same tests as in Hubert?

File: utils/check_repo.py [hunk 10]
```diff
@@ -122,6 +122,8 @@
     "TFRagTokenForGeneration",
     "Wav2Vec2ForCTC",
     "HubertForCTC",
+    "Wav2Vec2ForSequenceClassification",
```
Comment (patrickvonplaten) [inline utils/check_repo.py:125]: Yes! We have to discuss a bit with @Narsil how to best add those models to pipelines

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 11]
```diff
@@ -205,6 +202,9 @@ def __call__(
             padded_input_values = padded_inputs["input_values"]
             input_lengths = [padded_input_values.shape[-1] for _ in range(padded_input_values.shape[0])]
 
+        if isinstance(padded_inputs["input_values"][0], np.ndarray):
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:205]: My plan was to convert as early as possible from `np.float64` to `np.float32` here to save memory and speed, but we should look into making this better anyways! See: https://github.com/huggingface/transformers/issues/12862

File: src/transformers/models/hubert/modeling_hubert.py [hunk 12]
```diff
@@ -1070,3 +1072,106 @@ def forward(
...
+
+
+@add_start_docstrings(
+    """
+    Hubert Model with a sequence classification head on top (a linear layer over the pooled output) for tasks like
+    SUPERB.
+    """,
+    HUBERT_START_DOCSTRING,
+)
+class HubertForSequenceClassification(HubertPreTrainedModel):
```
Comment (patrickvonplaten) [inline src/transformers/models/hubert/modeling_hubert.py:1084]: Can we use a `# Copied from ...` here from Wav2Vec2 if the code is the same?

File: tests/test_modeling_hubert.py (hunk 8 above)
Comment (patrickvonplaten) [reply tests/test_modeling_hubert.py:660]: Let's just add a comment to the tests with high `atol` to be changed later once fairseq merged PR: https://github.com/pytorch/fairseq/pull/3572

File: src/transformers/models/hubert/convert_hubert_original_s3prl_checkpoint_to_pytorch.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/transformers/models/hubert/convert_hubert_original_s3prl_checkpoint_to_pytorch.py:45]: Let's check this a bit after evaluating the models

**Review by patrickvonplaten — COMMENTED** [review]
This PR looks to be in a very good shape already!
Before merging it would be great if we could:

- add a `#Copied from Wav2Vec2 ...` to the HuBERT code if it's 1-to-1 the same
- add one test per task for Wav2Vec2 as well 
- add at least one model for each task to either https://huggingface.co/superb or facebook (let's check with others here)
- run eval of the models on the datasets to check which models should be normalized and which shouldn't and adapt configs accordingly

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py (hunk 11 above)
Comment (anton-l) [reply src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:205]: I actually shifted it from `zero_mean_unit_var_norm()` to here just to fix the case when `do_normalize=False` and the arrays don't end up being converted to `float32`

File: src/transformers/models/hubert/modeling_hubert.py (hunk 12 above)
Comment (anton-l) [reply src/transformers/models/hubert/modeling_hubert.py:1084]: Probably, but I couldn't figure out how to make it work with `self.hubert = ` vs `self.wav2vec2 =`. That's the only difference other than model names.

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py (hunk 11 above)
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:205]: not sure this works tbh we have to be careful that normalization still works as expected with truncation and padding

File: utils/check_repo.py (hunk 10 above)
Comment (Narsil) [reply utils/check_repo.py:125]:
```
# Update this list for models that are not in any of the auto MODEL_XXX_MAPPING. Being in this list is an exception and
# should **not** be the rule.
```

Seems like the exception has grown quite a bit :)

File: src/transformers/models/hubert/modeling_hubert.py (hunk 12 above)
Comment (patrickvonplaten) [reply src/transformers/models/hubert/modeling_hubert.py:1084]: ah yeah you can have multiple name changes seperated by comma -> this should work:

```
# Copied from transformers.models.wav2vec2.modeling_wav2vec2.Wav2Vec2ForSequenceClassification with Wav2Vec2->Hubert, wav2vec2->hubert
```

Comment (patrickvonplaten) [reply src/transformers/models/hubert/modeling_hubert.py:1084]: See: https://github.com/huggingface/transformers/blob/3bb44662602b06ffd3ca25b585b5eb515bc22995/src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py#L2306

**Review by patrickvonplaten — APPROVED** [review]
